WordPress/gutenberg
# BorderControls: Alternate approach to handling popover positioning

Related:
- https://github.com/WordPress/gutenberg/pull/40836 

## What?

Provides opt-in sidebar positioning for `BorderControl` popovers without exposing or passing through popoverProps via the `BorderControl`'s public API.

## Why?

[New work refactoring the `Popover` ](https://github.com/WordPress/gutenberg/pull/40740) component requires the ability to position a `Popover` via props rather than a class name. 

This is an alternate approach to https://github.com/WordPress/gutenberg/pull/40836 after concerns about passing through `popoverProps` were [raised](https://github.com/WordPress/gutenberg/pull/40836#pullrequestreview-963198401).

## How?

- Fixes ColorPalette prop restructuring to avoid type error (https://github.com/WordPress/gutenberg/commit/362101992efc7195f498a6c9720ac8decb626c00)
- Removes popover positioning styles for the border block support from the block editor (https://github.com/WordPress/gutenberg/commit/495b61d5ac6336dd64789f2b23971d5dff01eff5)
- Replaces class name positioning of popovers with new approach (https://github.com/WordPress/gutenberg/commit/a8139732fff6c60721af6611c6d49868ed2203ff) 
    - If `__experimentalIsRenderedInSidebar` is `true` and `__experimentalSide` is set, the `BorderControl` will generate new styles and apply them internally via the inner Dropdown's `popoverProps`.
- Removes class name positioning of popovers from `BorderBoxControl` and updates it to leverage `BorderControl`'s new approach (https://github.com/WordPress/gutenberg/commit/3aea82f5535c6c552b96184ac71df8090c67dc50)
- Utilises `useResponsiveValue` to adjust when the sidebar positioning should occur (i.e. only on desktop).


## Testing Instructions

1. Run border control unit tests
    - `npm run test-unit packages/components/src/border-control/test/index.js`
    - `npm run test-unit packages/components/src/border-box-control/test/index.js`
2. Ensure no typing errors: `npm run build:package-types`
3. Load editor, create a post, add a group block, and select it
4. In the Inspector controls sidebar, toggle all the border color and style dropdowns for the border support control including both linked and unlinked views. Make sure the popovers are positioned correctly to the left of the sidebar.
5. Reduce the viewport size below medium, with the group block selected ensure the block settings are toggled on. 
6. Toggle on the various `BorderControl` popovers, they should all be the narrower default width and their positions not modified.

## Screenshots or screencast 

https://user-images.githubusercontent.com/60436221/167076084-fa7fd81e-8164-42ec-b1b8-906f6d64c4da.mp4
